src/types/dict.rs

```diff
@@ -308,27 +255,309 @@ impl PyDict {
+    where
+        K: ToPyObject,
+    {
+        fn inner(dict: &Py2<'_, PyDict>, key: Py2<'_, PyAny>) -> PyResult<()> {
+            err::error_on_minusone(dict.py(), unsafe {
+                ffi::PyDict_DelItem(dict.as_ptr(), key.as_ptr())
+            })
+        }
+
+        let py = self.py();
+        inner(self, key.to_object(py).attach_into(py))
+    }
+
+    fn keys(&self) -> Py2<'py, PyList> {
+        unsafe {
+            ffi::PyDict_Keys(self.as_ptr())
+                .assume_owned(self.py())
+                .downcast_into_unchecked()
+        }
+    }
+
+    fn values(&self) -> Py2<'py, PyList> {
+        unsafe {
+            ffi::PyDict_Values(self.as_ptr())
+                .assume_owned(self.py())
+                .downcast_into_unchecked()
+        }
+    }
+
+    fn items(&self) -> Py2<'py, PyList> {
+        unsafe {
+            ffi::PyDict_Items(self.as_ptr())
+                .assume_owned(self.py())
+                .downcast_into_unchecked()
+        }
+    }
+
+    fn iter(&self) -> PyDictIter<'py> {
+        PyDictIter::new(self.clone())
+    }
+
+    fn as_mapping(&self) -> &Py2<'py, PyMapping> {
+        unsafe { self.downcast_unchecked() }
+    }
+
+    fn update(&self, other: &Py2<'_, PyMapping>) -> PyResult<()> {
+        err::error_on_minusone(self.py(), unsafe {
+            ffi::PyDict_Update(self.as_ptr(), other.as_ptr())
+        })
+    }
+
+    fn update_if_missing(&self, other: &Py2<'_, PyMapping>) -> PyResult<()> {
+        err::error_on_minusone(self.py(), unsafe {
             ffi::PyDict_Merge(self.as_ptr(), other.as_ptr(), 0)
         })
     }
 }
 
+fn dict_len(dict: &Py2<'_, PyDict>) -> Py_ssize_t {
+    #[cfg(any(not(Py_3_8), PyPy, Py_LIMITED_API))]
+    unsafe {
+        ffi::PyDict_Size(dict.as_ptr())
+    }
+
+    #[cfg(all(Py_3_8, not(PyPy), not(Py_LIMITED_API)))]
+    unsafe {
+        (*dict.as_ptr().cast::<ffi::PyDictObject>()).ma_used
+    }
+}
+
+/// PyO3 implementation of an iterator for a Python `dict` object.
+pub struct PyDictIterator<'py>(PyDictIter<'py>);
+
+impl<'py> Iterator for PyDictIterator<'py> {
+    type Item = (&'py PyAny, &'py PyAny);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let (key, value) = self.0.next()?;
+        Some((key.into_gil_ref(), value.into_gil_ref()))
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<'py> ExactSizeIterator for PyDictIterator<'py> {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl<'a> std::iter::IntoIterator for &'a PyDict {
+    type Item = (&'a PyAny, &'a PyAny);
+    type IntoIter = PyDictIterator<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        PyDictIterator(PyDictIter::new(Py2::borrowed_from_gil_ref(&self).clone()))
+    }
+}
+
 /// PyO3 implementation of an iterator for a Python `dict` object.
-pub struct PyDictIterator<'py> {
-    dict: &'py PyDict,
+pub(crate) struct PyDictIter<'py> {
```

Comment:
For consistency, shouldn't this be `Py2DictIterator` or `PyDictIterator2`? Replacing `Iterator` by `Iterator` does not really explain the difference...